PICMI_Python/particles.py

```diff
@@ -152,18 +162,48 @@ class PICMI_GaussianBunchDistribution(_ClassWithInit):
       - centroid_velocity=[0,0,0]: Velocity (gamma*V) of the bunch centroid at t=0 (vector) [m/s]
       - velocity_divergence=[0,0,0]: Expansion rate of the bunch at t=0 (vector) [m/s/m]
     """
-    def __init__(self,n_physical_particles, rms_bunch_size,
-                 rms_velocity = [0.,0.,0.],
-                 centroid_position = [0.,0.,0.],
-                 centroid_velocity = [0.,0.,0.],
-                 velocity_divergence = [0.,0.,0.],
+    def __init__(self,n_physical_particles = None, density = None, rms_bunch_size,
+                 rms_velocity = None,
+                 centroid_position = None,
+                 centroid_velocity = None,
+                 velocity_divergence = None,
                  **kw):
+        
+        assert n_physical_particles is not None or density is not None, 'One of n_physical_particles or density must be speficied'
+
+        if n_physical_particles is None:
+            if rms_bunch_size.size==3:
+                n_physical_particles = np.prod(rms_bunch_size)*(np.pi*2)**1.5*density
+            if rms_bunch_size.size==2:
+                n_physical_particles = rms_bunch_size[0]**2*rms_bunch_size[1]*(np.pi*2)**1.5*density
+
+        if density is None:
+            if rms_bunch_size.size==3:
+                density = n_physica_particles/ (np.prod(rms_bunch_size)*(np.pi*2)**1.5)
+            if rms_bunch_size.size==2:
+                density = n_physical_particles/ (rms_bunch_size[0]**2*rms_bunch_size[1]*(np.pi*2)**1.5)
+
+
+
         self.n_physical_particles = n_physical_particles
         self.rms_bunch_size = rms_bunch_size
-        self.rms_velocity = rms_velocity
-        self.centroid_position = centroid_position
-        self.centroid_velocity = centroid_velocity
-        self.velocity_divergence = velocity_divergence
+        if(centroid_position is None):
```

Comment:
@s9105947 the question came up today how we would express such logic with the DataClasses.

Reply:
I'll jump in. See my comment above that the input parameters should always have a length of 3 (since this class should be independent of the dimensionality of the simulation). In that case, this logic is not needed.

But if something like this is needed, the dataclass allows the method `__post_init__` where this could be done. Type checking can still be done, ensuring that the input parameters are sequences.

Reply:
With respect to geometry, it's honestly also worth a thought to think about declaring everything 3D and expressing periodicity/symmetries separately for mapping. For instance, there are multiple ways to implement 1D, 2D and RZ, see for instance: https://doi.org/10.5281/zenodo.3266820 section 2.3.4

We do something similar for boosted frame simulations (express everything in the lab frame and convert in the background).

Reply:
> But if something like this is needed, the dataclass allows the method __post_init__ where this could be done. Type checking can still be done, ensuring that the input parameters are sequences.

I would say rather: the "Simulation" or "ParticleDistribution" class could implement `__init__` to overwrite options in the DataClass with more logic, when truly needed. No?